File: services/ContaPoupanca.php
<?php
require_once 'Conta.php';
class ContaPoupanca extends Conta
{
    public function depositar(float $valor): void
    {
        if ($valor <= 0) {
            throw new Exception("O valor do depósito deve ser positivo.");
        }

        $this->saldo += $valor;
    }

    public function sacar(float $valor): void
    {
        if ($valor <= 0) {
            throw new Exception("O valor do saque deve ser positivo.");
        }

        $valorComTaxa = $valor * 1.005;
        
        if ($valorComTaxa > $this->saldo) {
            throw new Exception("Saldo insuficiente para o saque.");
        }

        $this->saldo -= $valorComTaxa;
    }

    public function __toString(): string
    {
        return "Conta Poupança - " . parent::__toString();
    }
}

?>

File: model/ContaPoupanca.cs
using System;

namespace SistemaBancario
{
    public class ContaPoupanca : Conta
    {
        public ContaPoupanca(int numero, Titular titular) : base(numero, titular)
        {
        }
        public ContaPoupanca()
        {
        }
        public override void Depositar(decimal valor)
        {
            if (valor <= 0)
                throw new ArgumentException("O valor do depósito deve ser positivo.");

            this.saldo += valor;
        }

        public override void Sacar(decimal valor)
        {
            decimal valorComTaxa = valor * 1.005m; // Taxa de saque de 0.5%
            if (valor <= 0)
                throw new ArgumentException("O valor do saque deve ser positivo.");
            if (valorComTaxa > saldo)
                throw new InvalidOperationException("Saldo insuficiente para o saque.");

            saldo -= valorComTaxa;
        }

        public override string ToString()
        {
             return $"Conta Poupança - {base.ToString()}";
        }
    }
}